File: apps/api/app/services/ai/mock.py
from __future__ import annotations

import re
from typing import Any

PLATFORM_TO_CONTENT_TYPE = {
    "linkedin": "linkedin_post",
    "instagram": "instagram_caption",
    "x": "x_thread",
    "youtube": "youtube_script",
    "tiktok": "tiktok_script",
    "newsletter": "newsletter",
    "blog": "blog_outline",
}


def _to_reference(section: dict[str, Any]) -> dict[str, Any]:
    return {
        "sectionId": section["id"],
        "sectionTitle": section["section_title"],
        "sectionNumber": section["section_number"],
        "pageStart": section.get("page_start"),
        "pageEnd": section.get("page_end"),
    }


def _first_sentence(text: str, fallback: str) -> str:
    cleaned = re.sub(r"\s+", " ", text or "").strip()
    if not cleaned:
        return fallback
    match = re.match(r"^(.{20,220}?[.!?])\s", cleaned)
    if match:
        return match.group(1)
    return cleaned[:220]


def _chunk_sections(
    sections: list[dict[str, Any]],
    module_count: int,
) -> list[list[dict[str, Any]]]:
    count = max(1, min(module_count, len(sections)))
    groups: list[list[dict[str, Any]]] = [[] for _ in range(count)]
    for index, section in enumerate(sections):
        groups[index % count].append(section)
    return [group for group in groups if group]


def generate_course_outline(
    *,
    target_audience: str,
    course_objective: str,
    duration_label: str,
    module_count: int,
    difficulty_level: str,
    sections: list[dict[str, Any]],
) -> dict[str, Any]:
    if not sections:
        raise ValueError("Select at least one source section before generating.")

    groups = _chunk_sections(sections, module_count)
    all_refs = [_to_reference(section) for section in sections]
    primary = sections[0]

    modules = []
    for module_index, group in enumerate(groups):
        module_refs = [_to_reference(section) for section in group]
        lead = group[0]
        support = group[1:]
        lessons = [
            {
                "title": f"Understand: {lead['section_title']}",
                "summary": _first_sentence(
                    lead.get("extracted_text") or "",
                    f"Review the source material in {lead['section_title']}.",
                ),
                "learningObjectives": [
                    f"Explain the key idea from {lead['section_title']}.",
                    f"Identify how this idea serves {target_audience}.",
                ],
                "sourceReferences": [_to_reference(lead)],
            }
        ]
        for section in support[:2]:
            lessons.append(
                {
                    "title": f"Apply: {section['section_title']}",
                    "summary": _first_sentence(
                        section.get("extracted_text") or "",
                        f"Apply teaching points from {section['section_title']}.",
                    ),
                    "learningObjectives": [
                        f"Apply one practical takeaway from {section['section_title']}.",
                    ],
                    "sourceReferences": [_to_reference(section)],
                }
            )
        modules.append(
            {
                "title": f"Module {module_index + 1}: {lead['section_title']}",
                "description": _first_sentence(
                    lead.get("extracted_text") or "",
                    f"This module is grounded in {lead['section_title']}.",
                ),
                "lessons": lessons,
                "sourceReferences": module_refs,
            }
        )

    learning_outcomes = [
        f"Describe the core message drawn from {primary['section_title']}.",
        f"Apply the teaching points for {target_audience}.",
        "Connect selected source sections into a coherent learning path.",
    ]
    if difficulty_level == "advanced":
        learning_outcomes.append(
            "Critique and adapt the source ideas for advanced learners.",
        )

    quiz_suggestions = []
    for section in sections[:5]:
        snippet = _first_sentence(
            section.get("extracted_text") or "",
            section["section_title"],
        )
        quiz_suggestions.append(
            f"Based on {section['section_title']}: what is the main idea "
            f"behind “{snippet[:80]}”?"
        )

    return {
        "title": f"{primary['section_title']} Course",
        "description": " ".join(
            [
                f"A {difficulty_level} course for {target_audience}.",
                f"Objective: {course_objective}.",
                f"Duration focus: {duration_label}.",
                f"Content is structured only from the {len(sections)} "
                "selected source section(s).",
            ]
        ),
        "targetAudience": target_audience,
        "learningOutcomes": learning_outcomes,
        "modules": modules,
        "quizSuggestions": quiz_suggestions,
        "sourceReferences": all_refs,
        "groundingNotes": [
            "Generated by the mock AI provider using only selected source sections.",
            "No external facts were added beyond the creator inputs and source text.",
        ],
    }


def _label_platform(platform: str) -> str:
    return {
        "linkedin": "LinkedIn post",
        "instagram": "Instagram caption",
        "x": "X thread",
        "youtube": "YouTube script",
        "tiktok": "TikTok / Reel script",
        "newsletter": "Newsletter",
        "blog": "Blog outline",
    }.get(platform, "Content")


def _length_budget(length: str) -> int:
    if length == "short":
        return 280
    if length == "long":
        return 1200
    return 600


def _build_social_body(
    *,
    platform: str,
    tone: str,
    length: str,
    target_audience: str,
    call_to_action: str,
    section_title: str,
    idea: str,
    variant: int,
) -> str:
    budget = _length_budget(length)
    opener = idea

    if platform == "x":
        tweets = [
            f"1/{min(3, variant + 2)} For {target_audience}: {opener}",
            f"2/ Grounded in “{section_title}”. Tone: {tone}.",
            f"3/ {call_to_action}",
        ]
        return "\n\n".join(tweets)[: budget + 120]

    if platform in {"youtube", "tiktok"}:
        return "\n\n".join(
            [
                f"HOOK: {opener}",
                f"CONTEXT: This script is grounded in “{section_title}”.",
                (
                    f"BODY: Speak to {target_audience} in a {tone} voice. "
                    "Stay faithful to the source idea above — do not invent "
                    "extra claims."
                ),
                f"CTA: {call_to_action}",
            ]
        )[: budget + 200]

    if platform == "blog":
        return "\n".join(
            [
                f"Outline (variant {variant})",
                f"1. Introduction — {opener}",
                f"2. Core teaching from “{section_title}”",
                f"3. Practical application for {target_audience}",
                f"4. Closing CTA — {call_to_action}",
                "",
                "Notes: Keep every claim traceable to the selected source section.",
            ]
        )

    if platform == "newsletter":
        return "\n".join(
            [
                f"Subject angle: {section_title}",
                "",
                f"Hello {target_audience},",
                "",
                opener,
                "",
                f'This edition stays close to the source material in “{section_title}”.',
                "",
                call_to_action,
            ]
        )[: budget + 200]

    return "\n".join(
        [
            opener,
            "",
            f"Drawn from “{section_title}” for {target_audience}.",
            f"Voice: {tone}.",
            "",
            call_to_action,
        ]
    )[: budget + 80]


def generate_social_content(
    *,
    platform: str,
    tone: str,
    length: str,
    target_audience: str,
    call_to_action: str,
    output_count: int,
    sections: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    if not sections:
        raise ValueError("Select at least one source section before generating.")

    count = max(1, min(int(output_count), 5))
    outputs: list[dict[str, Any]] = []
    content_type = PLATFORM_TO_CONTENT_TYPE.get(platform, "linkedin_post")

    for index in range(count):
        section = sections[index % len(sections)]
        idea = _first_sentence(
            section.get("extracted_text") or "",
            f"Key idea from {section['section_title']}",
        )
        body = _build_social_body(
            platform=platform,
            tone=tone,
            length=length,
            target_audience=target_audience,
            call_to_action=call_to_action,
            section_title=section["section_title"],
            idea=idea,
            variant=index + 1,
        )
        outputs.append(
            {
                "contentType": content_type,
                "platform": platform,
                "title": (
                    f"{_label_platform(platform)} from "
                    f"{section['section_title']} (#{index + 1})"
                ),
                "body": body,
                "sourceReferences": [_to_reference(section)],
            }
        )
    return outputs

File: apps/web/lib/ai/providers/mock.ts
import { buildCourseOutlinePrompt } from "@/lib/ai/prompts/course-outline";
import { buildSocialContentPrompt } from "@/lib/ai/prompts/social-content";
import { buildVideoStoryboardPrompt } from "@/lib/ai/prompts/video-storyboard";
import {
  courseOutlineSchema,
  type CourseGenerationInput,
  type CourseGenerationSection,
  type CourseOutline,
  type SourceReference,
} from "@/lib/ai/schemas/course";
import {
  generatedSocialContentSchema,
  platformToContentType,
  type GeneratedSocialContent,
  type SocialGenerationInput,
  type SocialLength,
  type SocialPlatform,
} from "@/lib/ai/schemas/social";
import {
  videoStoryboardSchema,
  type VideoStoryboard,
  type VideoStoryboardInput,
} from "@/lib/ai/schemas/video";
import type { AIProvider } from "@/lib/ai/providers/types";

const MIN_SCENE_SECONDS = 2.5;
const MAX_SCENE_SECONDS = 7;
const MAX_SCENES_TEXT = 8;
const MAX_SCENES_FILM = 20;
const MAX_CAPTION_CHARS = 220;

function clampCaption(text: string): string {
  const clean = text.replace(/\s+/g, " ").trim();
  if (clean.length <= MAX_CAPTION_CHARS) return clean;
  return `${clean.slice(0, MAX_CAPTION_CHARS - 1).trimEnd()}…`;
}

function estimateSceneDuration(text: string): number {
  const words = text.trim().split(/\s+/).filter(Boolean).length;
  const seconds = words / 3; // ~3 words/second reading pace
  return Number(
    Math.min(MAX_SCENE_SECONDS, Math.max(MIN_SCENE_SECONDS, seconds)).toFixed(1),
  );
}

function splitSentences(text: string): string[] {
  return text
    .replace(/\s+/g, " ")
    .split(/(?<=[.!?])\s+/)
    .map((s) => s.trim())
    .filter(Boolean);
}

function toReference(section: CourseGenerationSection): SourceReference {
  return {
    sectionId: section.id,
    sectionTitle: section.sectionTitle,
    sectionNumber: section.sectionNumber,
    pageStart: section.pageStart,
    pageEnd: section.pageEnd,
  };
}

function firstSentence(text: string, fallback: string): string {
  const cleaned = text.replace(/\s+/g, " ").trim();
  if (!cleaned) return fallback;
  const match = cleaned.match(/^(.{20,220}?[.!?])\s/);
  if (match) return match[1];
  return cleaned.slice(0, 220);
}

function chunkSections(
  sections: CourseGenerationSection[],
  moduleCount: number,
): CourseGenerationSection[][] {
  const count = Math.max(1, Math.min(moduleCount, sections.length));
  const groups: CourseGenerationSection[][] = Array.from(
    { length: count },
    () => [],
  );

  sections.forEach((section, index) => {
    groups[index % count].push(section);
  });

  return groups.filter((group) => group.length > 0);
}

/**
 * Deterministic mock generator.
 * Grounds titles/summaries in selected section text and always attaches references.
 */
export class MockAIProvider implements AIProvider {
  async generateCourseOutline(
    input: CourseGenerationInput,
  ): Promise<CourseOutline> {
    // Keep prompt construction in the path so future providers share it.
    void buildCourseOutlinePrompt(input);

    if (input.sections.length === 0) {
      throw new Error("Select at least one source section before generating.");
    }

    const groups = chunkSections(input.sections, input.moduleCount);
    const allRefs = input.sections.map(toReference);
    const primary = input.sections[0];

    const modules = groups.map((group, moduleIndex) => {
      const moduleRefs = group.map(toReference);
      const lead = group[0];
      const support = group.slice(1);

      const lessons = [
        {
          title: `Understand: ${lead.sectionTitle}`,
          summary: firstSentence(
            lead.extractedText,
            `Review the source material in ${lead.sectionTitle}.`,
          ),
          learningObjectives: [
            `Explain the key idea from ${lead.sectionTitle}.`,
            `Identify how this idea serves ${input.targetAudience}.`,
          ],
          sourceReferences: [toReference(lead)],
        },
        ...support.slice(0, 2).map((section) => ({
          title: `Apply: ${section.sectionTitle}`,
          summary: firstSentence(
            section.extractedText,
            `Apply teaching points from ${section.sectionTitle}.`,
          ),
          learningObjectives: [
            `Apply one practical takeaway from ${section.sectionTitle}.`,
          ],
          sourceReferences: [toReference(section)],
        })),
      ];

      return {
        title: `Module ${moduleIndex + 1}: ${lead.sectionTitle}`,
        description: firstSentence(
          lead.extractedText,
          `This module is grounded in ${lead.sectionTitle}.`,
        ),
        lessons,
        sourceReferences: moduleRefs,
      };
    });

    const learningOutcomes = [
      `Describe the core message drawn from ${primary.sectionTitle}.`,
      `Apply the teaching points for ${input.targetAudience}.`,
      `Connect selected source sections into a coherent learning path.`,
    ];

    if (input.difficultyLevel === "advanced") {
      learningOutcomes.push(
        "Critique and adapt the source ideas for advanced learners.",
      );
    }

    const quizSuggestions = input.sections.slice(0, 5).map((section) => {
      const snippet = firstSentence(
        section.extractedText,
        section.sectionTitle,
      );
      return `Based on ${section.sectionTitle}: what is the main idea behind “${snippet.slice(0, 80)}”?`;
    });

    const outline = {
      title: `${primary.sectionTitle} Course`,
      description: [
        `A ${input.difficultyLevel} course for ${input.targetAudience}.`,
        `Objective: ${input.courseObjective}.`,
        `Duration focus: ${input.durationLabel}.`,
        `Content is structured only from the ${input.sections.length} selected source section(s).`,
      ].join(" "),
      targetAudience: input.targetAudience,
      learningOutcomes,
      modules,
      quizSuggestions,
      sourceReferences: allRefs,
      groundingNotes: [
        "Generated by the mock AI provider using only selected source sections.",
        "No external facts were added beyond the creator inputs and source text.",
      ],
    };

    return courseOutlineSchema.parse(outline);
  }

  async generateSocialContent(
    input: SocialGenerationInput,
  ): Promise<GeneratedSocialContent[]> {
    void buildSocialContentPrompt(input);

    if (input.sections.length === 0) {
      throw new Error("Select at least one source section before generating.");
    }

    const count = Math.max(1, Math.min(input.outputCount, 5));
    const outputs: GeneratedSocialContent[] = [];

    for (let index = 0; index < count; index += 1) {
      const section = input.sections[index % input.sections.length];
      const idea = firstSentence(
        section.extractedText,
        `Key idea from ${section.sectionTitle}`,
      );
      const body = buildMockSocialBody({
        platform: input.platform,
        tone: input.tone,
        length: input.length,
        targetAudience: input.targetAudience,
        callToAction: input.callToAction,
        sectionTitle: section.sectionTitle,
        idea,
        variant: index + 1,
      });

      outputs.push(
        generatedSocialContentSchema.parse({
          contentType: platformToContentType[input.platform],
          platform: input.platform,
          title: `${labelPlatform(input.platform)} from ${section.sectionTitle} (#${index + 1})`,
          body,
          sourceReferences: [toReference(section)],
        }),
      );
    }

    return outputs;
  }

  async generateVideoStoryboard(
    input: VideoStoryboardInput,
  ): Promise<VideoStoryboard> {
    // Keep prompt construction in the path so future providers share it.
    void buildVideoStoryboardPrompt(input);

    const trimmed = input.sourceText.trim();
    if (!trimmed) {
      throw new Error("Provide some text to generate a video from.");
    }

    const maxScenes =
      input.mode === "SCRIPT_TO_FILM" ? MAX_SCENES_FILM : MAX_SCENES_TEXT;

    let chunks: string[];
    if (input.mode === "SCRIPT_TO_FILM") {
      const paragraphs = trimmed
        .split(/\n\s*\n|\n(?=(?:INT\.|EXT\.|SCENE\b))/i)
        .map((p) => p.trim())
        .filter(Boolean);
      chunks = paragraphs.length > 1 ? paragraphs : splitSentences(trimmed);
    } else {
      chunks = splitSentences(trimmed);
      if (chunks.length === 0) chunks = [trimmed];
    }

    if (chunks.length === 0) chunks = [trimmed];

    // Merge neighbours if we exceed the scene budget so no text is dropped.
    if (chunks.length > maxScenes) {
      const groupSize = Math.ceil(chunks.length / maxScenes);
      const merged: string[] = [];
      for (let i = 0; i < chunks.length; i += groupSize) {
        merged.push(chunks.slice(i, i + groupSize).join(" "));
      }
      chunks = merged;
    }

    const scenes = chunks.map((chunk) => ({
      caption: clampCaption(chunk),
      durationSeconds: estimateSceneDuration(chunk),
    }));

    return videoStoryboardSchema.parse({ scenes });
  }
}

function labelPlatform(platform: SocialPlatform): string {
  switch (platform) {
    case "linkedin":
      return "LinkedIn post";
    case "instagram":
      return "Instagram caption";
    case "x":
      return "X thread";
    case "youtube":
      return "YouTube script";
    case "tiktok":
      return "TikTok / Reel script";
    case "newsletter":
      return "Newsletter";
    case "blog":
      return "Blog outline";
    default:
      return "Content";
  }
}

function lengthBudget(length: SocialLength): number {
  if (length === "short") return 280;
  if (length === "long") return 1200;
  return 600;
}

function buildMockSocialBody(input: {
  platform: SocialPlatform;
  tone: string;
  length: SocialLength;
  targetAudience: string;
  callToAction: string;
  sectionTitle: string;
  idea: string;
  variant: number;
}): string {
  const budget = lengthBudget(input.length);
  const opener = `${input.idea}`;

  if (input.platform === "x") {
    const tweets = [
      `1/${Math.min(3, input.variant + 2)} For ${input.targetAudience}: ${opener}`,
      `2/ Grounded in “${input.sectionTitle}”. Tone: ${input.tone}.`,
      `3/ ${input.callToAction}`,
    ];
    return tweets.join("\n\n").slice(0, budget + 120);
  }

  if (input.platform === "youtube" || input.platform === "tiktok") {
    return [
      `HOOK: ${opener}`,
      `CONTEXT: This script is grounded in “${input.sectionTitle}”.`,
      `BODY: Speak to ${input.targetAudience} in a ${input.tone} voice. Stay faithful to the source idea above — do not invent extra claims.`,
      `CTA: ${input.callToAction}`,
    ]
      .join("\n\n")
      .slice(0, budget + 200);
  }

  if (input.platform === "blog") {
    return [
      `Outline (variant ${input.variant})`,
      `1. Introduction — ${opener}`,
      `2. Core teaching from “${input.sectionTitle}”`,
      `3. Practical application for ${input.targetAudience}`,
      `4. Closing CTA — ${input.callToAction}`,
      ``,
      `Notes: Keep every claim traceable to the selected source section.`,
    ].join("\n");
  }

  if (input.platform === "newsletter") {
    return [
      `Subject angle: ${input.sectionTitle}`,
      ``,
      `Hello ${input.targetAudience},`,
      ``,
      opener,
      ``,
      `This edition stays close to the source material in “${input.sectionTitle}”.`,
      ``,
      input.callToAction,
    ]
      .join("\n")
      .slice(0, budget + 200);
  }

  // LinkedIn / Instagram default
  return [
    opener,
    ``,
    `Drawn from “${input.sectionTitle}” for ${input.targetAudience}.`,
    `Voice: ${input.tone}.`,
    ``,
    input.callToAction,
  ]
    .join("\n")
    .slice(0, budget + 80);
}

